### PR TITLE
Feature - daily updates

### PR DESCRIPTION
**Pull request by danielsp45**
### Achievements 
- [x] Daily updates feature added
- [x] Switch the daily updates messaging format from private message to admin channel
- [x] Make the bot send the daily update at a specific hour every 24 hours

#### Details
Every time a belt attribution is made, the information of that attribution such as the Ninja o received the belt, the date and who attributed it are stored in a json file. Every 24 hours the program searches in the json file for attributions that happened in the last 24 hours and sends it in a private message to all the admins. 
By mistake I made the bot send the updates as a private message to all the admins but I intend to switch it to just send it to a admin channel.

<details>

image

### CODE REVIEW

File: bot/cogs/belts.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            "- Apresentar o teu projeto a outros ninjas",
+            "- Saber o nome de 5 ninjas e 2 mentores",
+            "- Completar o Lightbot"],
+        "color" : "FFFFFF"
+        },
+    "Amarelo" : {
+        "objectives" : [
+            "- Estar presente em 4 sessões", 
+            "- Fazer um projeto e apresentar à mesa"],
+        "color" : "F7E129"
```
Comment (nelsonmestevao) [inline bot/cogs/belts.json:14]: I don't think it really matters but @laraujo7 choose as colours for each level the following:

- Branco : #e9e9e9
- Amarelo : #fcd767
- Azul : #7f9acd
- Verde : #09a777
- Laranja : #f79520
- Vermelho : #ec2027
- Roxo : #6f5977
- Preto : #383b3f

File: bot/client.py [hunk 2]
```diff
@@ -16,23 +17,28 @@
 async def on_ready():
     print(f'We have logged in as {client.user}')
 
+    DailyReport(client).report.start()
+    print("The task has been loaded")
```
Comment (nelsonmestevao) [inline bot/client.py:21]: I think we could improve our messages like this one using the [logging library](https://docs.python.org/3/library/logging.html). :bulb:

**Review by nelsonmestevao — COMMENTED** [review]
This is great work :clap: 

I think we should go another route in the way we store the events. Instead of using a file (in this case, a json file) to store all the events that happen, what do you think about using a more robust solution like sqlite3? It is simple in the sense that just a file (easy setup, deployment and backups) provides a battle tested way to read and write operations (like concurrent operations etc). 

There are libraries like [SQLAlchemy](https://www.sqlalchemy.org/) that provides easy ways to interact with databases.

File: bot/cogs/belts.json (hunk 1 above)
Comment (danielsp45) [reply bot/cogs/belts.json:14]: Nice, at the time I didn't know which color codes tu use. I'll put them in 💪🏻

File: bot/cogs/daily_report.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        if daily_logs != []:
+            for info in daily_logs:
+                msg += f'**{info["mentor_id"]}** atribuiu a **{info["ninja_id"]}** o cinturão {info["belt_attributed"]} \n'
+
+            embed = discord.Embed(
+                title = ":scroll: Histórico de atribuições",
+                description = msg
+            )
+            for member in admin_members:
+                await member.send(embed=embed)
```
[the remark this answers is not in the document]
Comment (danielsp45) [reply bot/cogs/daily_report.py:61]: Yes, it is required by discord.py

[the remark this answers is not in the document]
Comment (danielsp45) [reply bot/cogs/daily_report.py:61]: Yes, it is required by discord.py

File: bot/settings.py [hunk 4]
```diff
@@ -4,3 +4,4 @@
 load_dotenv()
 
 BOT_TOKEN = os.environ['BOT_TOKEN']
+GUILD_ID = 959927158745202728
```
[the remark this answers is not in the document]
Comment (danielsp45) [reply bot/settings.py:7]: You're right, I forgot to mention that I had to add that in order to the task work in the specific server and then forgot to do that. I'll change it as soon as possible

File: bot/client.py (hunk 2 above)
Comment (danielsp45) [reply bot/client.py:21]: Sounds like a really good idea!

**Comment by danielsp45** [comment]
> This is great work 👏
> 
> I think we should go another route in the way we store the events. Instead of using a file (in this case, a json file) to store all the events that happen, what do you think about using a more robust solution like sqlite3? It is simple in the sense that just a file (easy setup, deployment and backups) provides a battle tested way to read and write operations (like concurrent operations etc).
> 
> There are libraries like [SQLAlchemy](https://www.sqlalchemy.org/) that provides easy ways to interact with databases.

Yes, I agree with this. When using the json file I also got the sense that using a proper database would be a better idea. I'm going to work on the proposed changes and then jump right into this refactor 💪🏻

**Comment by nelsonmestevao** [comment]
> Yes, I agree with this. When using the json file I also got the sense that using a proper database would be a better idea. I'm going to work on the proposed changes and then jump right into this refactor 💪🏻

Let me know if you need any help :muscle:

**Comment by RuiL1904** [comment]
Great work, @danielsp45! 🙌🏼 
I've never used SQLAlchemy, but I would like to hear, from you @nelsonmestevao, what are the benefits on using it for this project. I once used [aiosqlite](https://github.com/omnilib/aiosqlite) - it's small and fast and on top of that it's asynchronous (which will combine very well with the discord asynchronous library). In a brief research, I've also found [asqlite](https://github.com/Rapptz/asqlite) which was made by Rapptz (the main `discord.py` maintainer).

**Comment by nelsonmestevao** [comment]
> I once used [aiosqlite](https://github.com/omnilib/aiosqlite) - it's small and fast and on top of that it's asynchronous (which will combine very well with the discord asynchronous library). In a brief research, I've also found [asqlite](https://github.com/Rapptz/asqlite) which was made by Rapptz (the main `discord.py` maintainer).

As far as I can tell, both suggestions are not much different from the standard library [sqlite3](https://docs.python.org/3/library/sqlite3.html). The thing that I most value from SQLAlchemy is the ORM (Object Relational Mapper) functionality that makes operations easier.

File: bot/cogs/daily_report.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, client):
+        self.client = client
+
+
+    @tasks.loop(seconds=10.0)
+    async def report(self):
+        daily_logs = []
+
+        guild = self.client.get_guild(GUILD_ID)                     # guild the bot is operating in 
+        admin_role = get_role_from_name(guild, "🛡️ Admin")          # guild  Admin role
```
Comment (nelsonmestevao) [inline bot/cogs/daily_report.py:34]: Should we create a `constansts.py` or even a roles Enum class with all the guild roles if we ever need to change them?

File: bot/client.py [hunk 6]
```diff
@@ -12,27 +16,56 @@
...
     )
 
 @client.command()
 async def unload(ctx, extension):
-    client.unload_extension(f'bot.cogs.{extension}')
+    '''
+    Command to unload an extension.
     
+    Parameters:
+    ctx (discord.ext.commands.Context): Represents the context in which a command is being invoked under.
```
Comment (nelsonmestevao) [inline bot/client.py:55]: I think we should move the type hinting to the function signature (that way we can get feedback from tools like `mypy` and IDEs) :bulb:

**Comment by danielsp45** [comment]
All of the features and fixes for this PR that are ready for review:
- Daily updates feature added
-  Switch the daily updates messaging format from private message to admin channel
-  Make the bot send the daily update at a specific hour every 24 hours
-  Switch the json file for SQLAlchemy to store all the logs
-  Implement loggings for debugging 

Me and @RuiL1904 plan to optimize the files organisation and to optimize the current Enum class in future PR's

**Review by nelsonmestevao — APPROVED** [review]
LGTM :rocket:

**Pull request merged**